### ISSUE TITLE
cosine calculation result > 1, when using HalfTensor vectors in pytorch

### ISSUE THREAD

**Issue by AaronHeee** [NVIDIA/apex#211]
When using `torch.nn.functional.cosine_similarity` to calculate the cosine similarity between vector `a` and `b`, we found the result is larger than 1 leading to errors in BCE loss calculation.

### Test vectors:
```
a = [ 8.9645e-04, 3.6120e-04, 4.9973e-04, -7.9775e-04, -1.6153e-04, -3.4118e-04,  1.4961e-04, -7.6830e-05,  5.6601e-04, -1.0824e-03, 3.1352e-04,  1.1563e-04,  3.1352e-04, -4.0674e-04,  3.1805e-04, 1.1301e-04,  8.6737e-04, -3.1304e-04, -3.1996e-04, -3.2854e-04, 1.0891e-03,  1.0300e-03, -2.9445e-05,  1.7333e-04, -6.3276e-04, -7.7724e-04, -4.6551e-05, -1.2627e-03, -1.7881e-05, -2.7943e-04, -3.6240e-04, -4.0603e-04, -1.8287e-04, -3.1400e-04,  7.3338e-04, -6.1226e-04, -4.3726e-04,  6.7186e-04, -6.5863e-05,  1.0643e-03, -2.5225e-04,  3.5143e-04,  8.7881e-04, -7.8583e-04,  3.1924e-04, -7.3969e-05,  8.0347e-04, -2.5821e-04,  5.4455e-04,  2.0814e-04, -4.9925e-04, -5.0306e-04,  1.3971e-04,  4.3273e-04, -1.0929e-03, -2.2650e-05, -3.7003e-04, -4.6825e-04,  7.3433e-04,  2.3901e-04, -9.1612e-05, -1.8525e-04, -9.3508e-04,  1.0853e-03, -1.8895e-05, -8.7023e-04,  6.3276e-04,  2.4247e-04,  2.4343e-04,  1.3697e-04, -3.5548e-04, -6.0177e-04,  1.3471e-05, -1.8442e-04, -2.2340e-04, -4.8018e-04,  4.0984e-04,  1.3323e-03,  1.5891e-04, -1.9002e-04, -3.2973e-04, -3.0255e-04,  3.1114e-04,  4.2105e-04,  4.7231e-04, -1.1082e-03, -9.1839e-04, -3.6335e-04, -2.3293e-04,  3.6597e-04, 2.7800e-04,  4.0233e-05,  9.1600e-04, -3.4451e-04, -5.1117e-04, 2.9230e-04, -8.9359e-04,  3.5977e-04, -1.6105e-04,  4.8828e-04, 3.6955e-06,  7.2539e-05,  3.4213e-04,  5.6314e-04, -2.3234e-04, 2.1219e-04,  3.4761e-04,  4.3631e-04,  1.6308e-04, -7.6437e-04, -4.5228e-04, -2.8110e-04, -1.0138e-03,  1.0986e-03,  5.0116e-04, 1.7571e-04, -2.4319e-04, -1.6522e-04,  6.9678e-05, -4.6754e-04, -1.0347e-03, -9.2125e-04, -8.0872e-04, -1.2243e-04, -1.5318e-04, -3.9220e-04,  6.3753e-04, -4.1556e-04]
b = [ 9.0790e-04, 3.5095e-04, 4.8542e-04, -7.8535e-04, -1.6344e-04, -3.4213e-04,  1.3781e-04, -5.8532e-05,  5.8365e-04, -1.0777e-03, 3.1209e-04,  1.0413e-04,  3.0494e-04, -4.0865e-04,  3.0541e-04, 1.2112e-04,  8.7786e-04, -3.1948e-04, -3.1066e-04, -3.0923e-04, 1.0996e-03,  1.0290e-03, -2.9206e-06,  1.5795e-04, -6.3324e-04, -7.8869e-04, -4.5180e-05, -1.2636e-03, -2.6822e-06, -2.8658e-04, -3.6812e-04, -3.9124e-04, -1.6165e-04, -3.0184e-04,  7.2289e-04, -6.0749e-04, -4.3988e-04,  6.7043e-04, -5.2154e-05,  1.0433e-03, -2.3961e-04,  3.5930e-04,  8.8024e-04, -8.0204e-04,  3.2306e-04, -7.0512e-05,  8.1444e-04, -2.5868e-04,  5.4407e-04,  2.0802e-04, -4.9353e-04, -4.7541e-04,  1.4818e-04,  4.2629e-04, -1.0691e-03, -2.6643e-05, -3.4928e-04, -4.5252e-04,  7.1096e-04,  2.7347e-04, -1.0026e-04, -2.1195e-04, -9.2554e-04,  1.0729e-03, -4.9472e-05, -8.8930e-04,  6.3038e-04,  2.6751e-04,  2.6464e-04,  1.6642e-04, -3.7622e-04, -5.9319e-04, -1.4067e-05, -1.8919e-04, -2.4235e-04, -4.5562e-04,  3.9268e-04,  1.3170e-03,  1.5259e-04, -2.0432e-04, -3.4261e-04, -3.0041e-04,  3.1757e-04,  4.2605e-04,  4.5037e-04, -1.1177e-03, -9.2554e-04, -3.7646e-04, -2.3937e-04,  3.7217e-04, 2.8610e-04,  2.9147e-05,  9.0933e-04, -3.7169e-04, -4.8494e-04, 2.8896e-04, -8.9693e-04,  3.4690e-04, -1.5604e-04,  4.9114e-04, 1.6093e-06,  9.2566e-05,  3.3236e-04,  5.6982e-04, -2.6822e-04, 2.2578e-04,  3.5024e-04,  4.4060e-04,  1.5271e-04, -7.7486e-04, -4.5347e-04, -2.7037e-04, -9.8324e-04,  1.1024e-03,  5.0735e-04, 1.4532e-04, -2.5368e-04, -1.4329e-04,  6.3419e-05, -4.5323e-04, -1.0300e-03, -9.0981e-04, -8.1730e-04, -9.7871e-05, -1.5974e-04, -3.9220e-04,  6.6566e-04, -4.0913e-04]
```

### Result on torch.FloatTensor:
`torch.nn.functional.cosine_similarity(a, b) ` = 0.9996

### Result on torch.HalfTensor:
`torch.nn.functional.cosine_similarity(a.half(), b.half()) ` = 1.0020

Therefore, the cosine similarity is larger than 1, which is harmful for the further loss function calculation. Is there any way to relieve this problem, except converting back to FloatTensor?
Thanks a lot.

**Comment by mcarilli** [comment]
Seems to me like this is an operation that should definitely be carried out in float.  I'm about to travel but I will try putting this on the FP32 lists early next week.

**Comment by ngimel** [comment]
Note that even computing cosine similarity in float sometimes resulted in values bigger than 1 https://github.com/pytorch/pytorch/issues/18057, it was fixed upstream only recently (see PR links in the upstream issue).
